go/enclave/rollupchain/rollup_chain.go

```diff
@@ -393,82 +403,91 @@ func (rc *RollupChain) produceBlockSubmissionResponse(block *types.Block, l2Head
 	}, nil
 }
 
-// Calculates and stores the block state, receipts and logs for the given block.
-func (rc *RollupChain) updateHeads(block *types.Block) (*common.L2RootHash, bool, error) {
+// Updates the heads of the L1 and L2 chains.
+func (rc *RollupChain) updateHeads(block *types.Block) (*common.L2RootHash, error) {
```

Comment:
This is where I made most of the changes. Previously, `handleGenesisRollup` was called on all blocks, and you had to infer from the return values whether it was a pre-genesis block, or a genesis block, or a post-genesis block. The flow was very confusing. Now, we determine the block's type, then have it handled by the corresponding switch branch.